plugins/analysis/common/src/main/kotlin/arrow/meta/plugins/analysis/phases/analysis/solver/collect/ConstraintCollection.kt

```diff
@@ -323,24 +325,74 @@ private fun SolverState.addConstraints(
   postConstraints: ArrayList<NamedConstraint>,
   bindingContext: ResolutionContext
 ) {
-  val lawSubject = findDescriptorFromRemoteLaw(descriptor) ?: findDescriptorFromLocalLaw(descriptor, bindingContext)
+  val lawSubject =
+    findDescriptorFromRemoteLaw(descriptor, bindingContext)
+    ?: findDescriptorFromLocalLaw(descriptor, bindingContext)
+  if (lawSubject is CallableDescriptor && lawSubject.fqNameSafe == FqName("arrow.analysis.post"))
+    throw Exception("trying to attach to post, this is wrong!")
   if (lawSubject != null) {
     val renamed = solver.renameConditions(DeclarationConstraints(descriptor, preConstraints, postConstraints), lawSubject)
     callableConstraints.add(renamed.descriptor, ArrayList(renamed.pre), ArrayList(renamed.post))
   }
   callableConstraints.add(descriptor, preConstraints, postConstraints)
 }
 
-private fun findDescriptorFromRemoteLaw(descriptor: DeclarationDescriptor): DeclarationDescriptor? =
+private fun findDescriptorFromRemoteLaw(
+  descriptor: DeclarationDescriptor,
+  context: ResolutionContext
+): DeclarationDescriptor? =
   descriptor.annotations().findAnnotation(FqName("arrow.analysis.Subject"))?.let { lawSubject ->
-    (lawSubject.argumentValueAsString("fqName"))
-      ?.let { FqName(it) }
-      ?.let { subjectFqName ->
-        val pck = subjectFqName.name.substringBeforeLast(".")
-        val fn = subjectFqName.name.split(".").lastOrNull()
-        descriptor.module.getPackage(pck)?.getMemberScope()?.getContributedDescriptors { it == fn }?.firstOrNull()
+    val name = lawSubject.argumentValueAsString("fqName")
+    val result = name?.let {
+      descriptor.module.obtainDeclaration(FqName(it), descriptor)
+    }
+    result.also {
+      if (it == null) {
+        if (name == null)
+          throw Exception(ErrorMessages.Parsing.subjectWithoutName(descriptor.fqNameSafe.name))
```

Comment:
should we report these instead of throwing an exception?

Reply:
There are two reasons:
- at some point these should go, they are really cases which should not happen, unless somebody tinkers with the annotations;
- there is no particular `PsiElement` to which they can be hung, so it was not clear how to turn those into messages.